### PR TITLE
UTY-1106: replace entity index with entity

### PR DESCRIPTION
**Pull request by jessicafalk**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
We have been using the entity index instead of the entity to identify entities in our game object representation module. This is not unique (you need the index and the version) and caused some interesting bugs.

 #### Tests
Ran it locally and checked whether the error still appeared. Now you can build out the gamelogic worker and connect/disconnect the client as often as you want without it throwing errors.

#### Documentation
internal implementation

#### Primary reviewers
@JonasImprobable @zeroZshadow

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs [hunk 1]
```diff
@@ -166,16 +166,16 @@ public ActivationManagerNotFoundException(string message) : base(message)
...
-            if (entityIndexToActivationManager.ContainsKey(entity.Index))
+            if (entityIndexToActivationManager.ContainsKey(entity))
             {
                 throw new ActivationManagerAlreadyExistsException($"MonoBehaviourActivationManager already exists for entityIndex {entity.Index}.");
             }
 
             var gameObject = EntityManager.GetComponentObject<GameObjectReference>(entity).GameObject;
             var store = new InjectableStore();
-            entityIndexToReaderWriterStore[entity.Index] = store;
+            entityIndexToReaderWriterStore[entity] = store;
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs]: Please change these to `.Add` instead

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectComponentDispatcherBase.cs [hunk 2]
```diff
@@ -24,20 +24,20 @@ internal abstract class GameObjectComponentDispatcherBase
         public ComponentGroup[] CommandRequestsComponentGroups { get; set; }
 
         public abstract void MarkComponentsAddedForActivation
-            (Dictionary<int, MonoBehaviourActivationManager> entityIndexToManagers);
+            (Dictionary<Entity, MonoBehaviourActivationManager> entityIndexToManagers);
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectComponentDispatcherBase.cs]: Can  you rename the argument name `entityIndexToManagers` to something like `entityToManager`?

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectComponentDispatcherBase.cs [hunk 3]
```diff
@@ -24,20 +24,20 @@ internal abstract class GameObjectComponentDispatcherBase
...
 
-        public abstract void InvokeOnComponentUpdateCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore);
-        public abstract void InvokeOnEventCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore);
-        public abstract void InvokeOnCommandRequestCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore);
-        public abstract void InvokeOnCommandResponseCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore);
-        public abstract void InvokeOnAuthorityGainedCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore);
-        public abstract void InvokeOnAuthorityLostCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore);
-        public abstract void InvokeOnAuthorityLossImminentCallbacks(Dictionary<int, InjectableStore> entityIndexToInjectableStore);
+        public abstract void InvokeOnComponentUpdateCallbacks(Dictionary<Entity, InjectableStore> entityIndexToInjectableStore);
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectComponentDispatcherBase.cs]: Same as above.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs [hunk 4]
```diff
@@ -12,26 +12,26 @@ namespace Improbable.Gdk.Core.GameObjectRepresentation
...
         private ILogDispatcher logger;
 
-        internal void RemoveActivationManagerAndReaderWriterStore(int entityIndex)
+        internal void RemoveActivationManagerAndReaderWriterStore(Entity entity)
         {
-            if (!entityIndexToActivationManager.ContainsKey(entityIndex))
+            if (!entityToActivationManager.ContainsKey(entity))
             {
-                throw new ActivationManagerNotFoundException($"MonoBehaviourActivationManager not found for entityIndex {entityIndex}.");
+                throw new ActivationManagerNotFoundException($"MonoBehaviourActivationManager not found for entity {entity.Index}.");
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs]: Nit: Could we print the `entity.Version` here too? Not sure if that's ever going to be useful. Just making sure we stay safe.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs [hunk 5]
```diff
@@ -166,16 +166,16 @@ public ActivationManagerNotFoundException(string message) : base(message)
 
         public void CreateActivationManagerAndReaderWriterStore(Entity entity)
         {
-            if (entityIndexToActivationManager.ContainsKey(entity.Index))
+            if (entityToActivationManager.ContainsKey(entity))
             {
-                throw new ActivationManagerAlreadyExistsException($"MonoBehaviourActivationManager already exists for entityIndex {entity.Index}.");
+                throw new ActivationManagerAlreadyExistsException($"MonoBehaviourActivationManager already exists for entity {entity.Index}.");
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs]: Same as above.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt [hunk 6]
```diff
@@ -78,7 +78,7 @@ namespace <#= qualifiedNamespace #>
             private static readonly InjectableId commandRequestHandlerInjectableId = new InjectableId(InjectableType.CommandRequestHandler, componentId);
 <# } #>
 
-            public override void MarkComponentsAddedForActivation(Dictionary<int, MonoBehaviourActivationManager> entityIndexToManagers)
+            public override void MarkComponentsAddedForActivation(Dictionary<Unity.Entities.Entity, MonoBehaviourActivationManager> entityToManagers)
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt]: Rename `entityToManagers` to `entityToManager`.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt [hunk 7]
```diff
@@ -88,12 +88,12 @@ namespace <#= qualifiedNamespace #>
...
                 for (var i = 0; i < entities.Length; i++)
                 {
-                    var activationManager = entityIndexToManagers[entities[i].Index];
+                    var activationManager = entityToManagers[entities[i]];
                     activationManager.AddComponent(componentId);
                 }
             }
 
-            public override void MarkComponentsRemovedForDeactivation(Dictionary<int, MonoBehaviourActivationManager> entityIndexToManagers)
+            public override void MarkComponentsRemovedForDeactivation(Dictionary<Unity.Entities.Entity, MonoBehaviourActivationManager> entityToManagers)
```
Comment (JonasImprobable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityGameObjectComponentDispatcherGenerator.tt]: Same as above.

**Comment by JonasImprobable** [comment]
Whoops, looks like I commented after this was merged. The concerns I made are all nits so things are fine the way they are right now.

**Pull request merged**